File: python/day3/part2.py
import re

in_date = ""

with open("input.txt") as f:
    in_date = f.read()

pattern = r"(don\'t\(\))|(do\(\))|(mul\((\d{1,3},\d{1,3})\))"

def mul(xy) -> int:
    x, y = (int(d) for d in xy.split(","))
    return x * y


carry_on = True
total = 0
for catch in re.finditer(pattern, in_date):
    if carry_on and catch.group(3):
        total += mul(catch[4])
    elif catch.group(2):
        carry_on = True
    elif catch.group(1):
        carry_on = False

print(total)

File: rust/day3/src/bin/part2.rs
use regex::Regex;

fn mul(xy: &str) -> i32 {
    let col: Vec<&str> = xy.split(',').collect();
    col[0].to_string().parse::<i32>().unwrap_or(0) * col[1].to_string().parse::<i32>().unwrap_or(0)
}

fn main() {
    // let txt = "xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))";
    let in_data = include_str!("../input.txt");

    let pattern = r"do\(\)|don\'t\(\)|mul\(\d{1,3},\d{1,3}\)";
    let re = Regex::new(pattern).unwrap();

    // let dates: Vec<&str> = re.find_iter(txt).map(|m| m.as_str()).collect();
    let mut total = 0;
    let mut carry_on = true;
    for item in re.find_iter(in_data) {
        let current = item.as_str();
        if current == "do()" {
            carry_on = true;
        } else if current == "don't()" {
            carry_on = false;
        }

        let p = r"\d{1,3},\d{1,3}";
        let subre = Regex::new(p).unwrap();
        let res = subre.find(&current);

        if carry_on && res.is_some() {
            total += mul(res.unwrap().as_str());
        }
    }
    println!("{}", total)
}
